pkg/licensing/scanner.go

```diff
@@ -22,21 +23,46 @@ func NewScanner(categories map[types.LicenseCategory][]string) Scanner {
 }
 
 func (s *Scanner) Scan(licenseName string) (types.LicenseCategory, string) {
-	normalized := NormalizeLicense(expression.SimpleExpr{License: licenseName})
-	var normalizedName string
-	switch normalized := normalized.(type) {
-	case expression.SimpleExpr:
-		normalizedName = normalized.License
-	case expression.CompoundExpr:
-		normalizedName = normalized.String()
-	}
+	visited := make(map[string]types.LicenseCategory)
+	category := s.traverseLicenseExpression(licenseName, visited)
 
-	for category, names := range s.categories {
-		if slices.Contains(names, normalizedName) {
-			return category, categoryToSeverity(category).String()
+	return category, categoryToSeverity(category).String()
+}
+
+// traverseLicenseExpression recursive parses license expression to detect correct license category:
+// For Simple Expression - use category of license
+// For Compound Expression:
+//   - `AND` operator - use category with maximum severity
+//   - `OR` operator - use category with minimum severity
+//   - one of expression has `UNKNOWN` category - use `UNKNOWN` category
+func (s *Scanner) traverseLicenseExpression(licenseName string, visited map[string]types.LicenseCategory) types.LicenseCategory {
+	category := types.CategoryUnknown
+
+	detectCategoryAndSeverity := func(expr expression.Expression) expression.Expression {
```

Comment:
nit:
The severity in the name is taken from the previous function name.
This function does not currently define severity. We will do that later.
```suggestion
	detectCategory := func(expr expression.Expression) expression.Expression {
```

Reply:
Good catch, that is indeed a bad name given the changes we've made.